File: screen-switcher-sample/src/main/java/com/jaynewstrom/screenswitchersample/ApplicationModule.java
package com.jaynewstrom.screenswitchersample;

import com.jaynewstrom.screenswitchersample.base.LeakWatcher;
import com.squareup.leakcanary.RefWatcher;

import dagger.Module;
import dagger.Provides;

@Module
final class ApplicationModule {
    private final ScreenSwitcherApplication application;
    private final RefWatcher refWatcher;

    ApplicationModule(ScreenSwitcherApplication application, RefWatcher refWatcher) {
        this.application = application;
        this.refWatcher = refWatcher;
    }

    @Provides LeakWatcher provideLeakWatcher() {
        return new LeakWatcher() {
            @Override public void watch(Object object) {
                refWatcher.watch(object);
            }
        };
    }
}

File: screen-switcher-sample/src/main/java/com/jaynewstrom/screenswitchersample/application/ApplicationModule.kt
package com.jaynewstrom.screenswitchersample.application

import android.content.Context
import com.jaynewstrom.screenswitchersample.core.LeakWatcher
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
internal class ApplicationModule(private val applicationContext: Context) {
    @Provides fun provideApplicationContext() = applicationContext

    @Module
    companion object {
        @Provides @JvmStatic @Singleton fun provideLeakWatcher(context: Context) = object : LeakWatcher {
            override fun <T> watch(t: T, description: String) {
                ScreenSwitcherApplication.watchObject(context, t, description)
            }
        }
    }
}
